pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,419 @@
+	postfixReceiveAdapter = "rcvadptr"
+
+	// watchTimeout is the timeout that the feedlet will wait for the Receiver Adapter to get a domain name.
+	watchTimeout = 5 * time.Minute
+
+	// secretNameKey is the name of the secret that contains the GitHub credentials.
+	secretNameKey = "secretName"
+	// secretKeyKey is the name of key inside the secret that contains the GitHub credentials.
+	secretKeyKey = "secretKey"
+)
+
+type GithubEventSource struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+	// servingclientset is a clientset for serving API groups
+	servingclientset servingclientset.Interface
+	// namespace where the feed is created
+	feedNamespace string
+	// serviceAccount that the container runs as. Launches Receive Adapter with the
+	// same Service Account
+	feedServiceAccountName string
+	// image for the receive adapter
+	image string
+}
+
+func NewGithubEventSource(kubeclientset kubernetes.Interface, servingclientset servingclientset.Interface, feedNamespace, feedServiceAccountName, image string) sources.EventSource {
+	return &GithubEventSource{
+		kubeclientset:          kubeclientset,
+		servingclientset:       servingclientset,
+		feedNamespace:          feedNamespace,
+		feedServiceAccountName: feedServiceAccountName,
+		image: image,
+	}
+}
+
+// TODO(n3wscott): Add a timeout for StopFeed.
+func (t *GithubEventSource) StopFeed(trigger sources.EventTrigger, feedContext sources.FeedContext) error {
+	log.Printf("stopping github webhook feed with context %+v", feedContext)
+
+	return t.deleteWebhook(trigger, feedContext)
+}
+
+// TODO(n3wscott): Add a timeout for StartFeed.
+func (t *GithubEventSource) StartFeed(trigger sources.EventTrigger, target string) (*sources.FeedContext, error) {
+
+	// Create the Receive Adapter Service that will accept incoming requests from GitHub.
+	service, err := t.createReceiveAdapter(trigger, target)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %v", err)
+	}
+
+	// TODO(n3wscott): look into using spew.
+	log.Printf("created Service: %+v", service)
+
+	// Start watching the Receive Adapter Service for it's updated domain name. This will be passed
+	// to GitHub as part of the webhook registration.
+	receiveAdaptorDomain, err := t.waitForServiceDomain(service.GetObjectMeta().GetName())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the service: %v", err)
+	}
+
+	return t.createWebhook(trigger, service.GetObjectMeta().GetName(), receiveAdaptorDomain)
+}
+
+func (t *GithubEventSource) waitForServiceDomain(serviceName string) (string, error) {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	wt := int64(watchTimeout / time.Second)
+	opts := metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: "v1alpha1",
+		},
+		FieldSelector:  fmt.Sprintf("metadata.name=%s", serviceName),
+		LabelSelector:  "receive-adapter=github",
```

Comment:
My thinking was that there might be many sources, so I am filtering with the label, and then of the sources that are github, I am using the field selector for the exact name.

Reply:
I am not sure this makes k8s faster, but it does prevent me from selecting on a service with the same name that is not the github RA.

Reply:
I understand that but I want to avoid the case where a name conflicts and the type is not a match. So I am trying to match on receive-adapter type AND name

Reply:
I think the client created at https://github.com/knative/eventing/pull/276/files#diff-15c4ba004c85a3abc13efbe47a731b58R121 will filter out anything that is not the expected type.

I think the case @n3wscott is concerned about is:
1. User creates Service A.
2. User creates a Feed that will also try to create service A.
3. `createReceiveAdapter` notices the existing service and does not create another one.
4. The watch watches the existing service which may not be suitable.

I'd argue that the issue here is actually step 3: `createReceiveAdapter` adopts an existing service without checking whether it's suitable. The suitability check (looking for the `receive-adapter` label) should happen there. Then by the time we get to the watch, we know for sure that the service being watched is correct.